src/msw/srchctrl.cpp

```diff
@@ -0,0 +1,448 @@
+
+        case WM_NCHITTEST:
+            {
+                POINT pt;
+                pt.x = GET_X_LPARAM(lParam);
+                pt.y = GET_Y_LPARAM(lParam);
+                ::MapWindowPoints(NULL, GetHwnd(), &pt, 1);
+
+                const wxPoint point = wxPoint(pt.x + bitmapWidth, pt.y);
+
+                if ( IsCancelButtonVisible() && m_cancelButtonRect.Contains(point) )
+                {
+                    m_mouseInCancelButton = true;
+                    *rc = HTCAPTION;
+                    processed = true;
+
+                    TRACKMOUSEEVENT tme = { sizeof(tme), TME_NONCLIENT | TME_LEAVE, GetHwnd() };
+                    TrackMouseEvent(&tme);
+                }
+                else if ( m_searchButtonRect.Contains(point) )
+                {
+                    *rc = HasMenu() ? HTCAPTION : HTCLIENT;
+                    processed = true;
+                }
+            }
+            break;
+
+        case WM_NCLBUTTONDOWN:
+            {
+                ClickButton();
+                processed = true;
+            }
+            break;
+
+        case WM_NCMOUSELEAVE:
+            {
+                m_mouseInCancelButton = false;
+            }
+            wxFALLTHROUGH;
+
+        case WM_NCMOUSEMOVE:
+            {
+                HighlightButton();
+            }
+            break;
+
+            // Reduce noticeable flicker by processing WM_MOUSEMOVE here
+            // by doing nothing if it just a mouse move only.
+        case WM_MOUSEMOVE:
+            {
+                // ...text can be selected though.
+                if ( wParam & MK_LBUTTON )
+                    break;
+            }
+            wxFALLTHROUGH;
+
+        case WM_NCLBUTTONDBLCLK:
+            {
+                *rc = 0;
+                processed = true;
+            }
+            break;
+    }
+
+    return processed || wxTextCtrl::MSWHandleMessage(rc, nMsg, wParam, lParam);
+}
+
+void wxSearchCtrl::DrawButtons(int width)
+{
+    const double xWidth = 80.;
+    const double radius = xWidth / 4. + 5;
+    const double scale  = 0.6 * (width / xWidth);
+    const double shift  = (width / 4.) - 1;
+
+    const wxColour bg = GetBackgroundColour();
+    const wxColour fg = GetForegroundColour().ChangeLightness(130);
+
+    wxWindowDC winDC(this);
+
+    {
+        wxGCDC gdc(winDC);
+        wxDCClipper clip(gdc, m_searchButtonRect);
+
+        gdc.SetDeviceOrigin(m_searchButtonRect.x + shift, shift);
+
+        gdc.SetBackground(bg);
+        gdc.Clear();
+
+        gdc.SetUserScale(scale, scale);
+
+        gdc.SetPen(wxPen(fg, 8));
+        gdc.SetBrush(*wxTRANSPARENT_BRUSH);
+
+        gdc.DrawCircle(xWidth - radius, radius, radius);
+        gdc.DrawLine(0, xWidth, xWidth/2. - 5, xWidth/2. + 5);
+    }
+
+    if ( IsCancelButtonVisible() )
+    {
+        wxGCDC gdc(winDC);
+        wxDCClipper clip(gdc, m_cancelButtonRect);
+
+        const double adj = GetLayoutDirection() == wxLayout_LeftToRight ? 1 : 2;
```

Comment:
I feel like this would merit a comment because it's not obvious why do we need to do this.